torch/distributed/_shard/api.py

```diff
@@ -118,3 +123,80 @@ def shard_parameter(
 
     # Now we can set the attribute appropriately.
     setattr(module, param_name, st)
+
+
+class ShardedModuleSwapper(abc.ABC):
+    @abc.abstractmethod
+    def process(self, module: nn.Module) -> nn.Module:
+        """
+        Processes a module and if needed swaps it with a custom sharded
+        Implementation. Should return ``None`` if no swapping should be
+        performed.
+
+        The passed in module would consist of ShardedTensors and a common way
```

Comment:
Maybe elaborate a bit more in terms of where the ShardedTensors come from. The sharder would produce ShardedTensors for modules based on the ShardingPlan and then call the ShardedModuleSwapper.